[HUDI-2069] Refactored String constants

sbernauer opened the pull request:
## What is the purpose of the pull request
Follow-Up of https://github.com/apache/hudi/pull/3111 adressing the feedback
Revert a change from https://github.com/apache/hudi/pull/3092 that broke the feature again by not passing the needed properties to kafka-client

## Brief change log
* Refactored String constants
* Pass the `hoodie.deltastreamer.source.kafka.value.deserializer.*` configs to the kafka clienr
## Verify this pull request

This pull request is a trivial rework / code cleanup without any test coverage.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

sbernauer commented:
@hudi-bot run azure

nsivabalan commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -381,7 +381,6 @@ object DataSourceWriteOptions {
-  // Schema to be used in custom kakfa deserializer
-  val KAFKA_AVRO_VALUE_DESERIALIZER_SCHEMA = "hoodie.deltastreamer.source.kafka.value.deserializer.schema"
+  val KAFKA_AVRO_VALUE_DESERIALIZER_OPT_KEY = "hoodie.deltastreamer.source.kafka.value.deserializer.class"
+  val DEFAULT_KAFKA_AVRO_VALUE_DESERIALIZER_OPT_VAL = "org.apache.hudi.utilities.deser.KafkaAvroSchemaDeserializer"
```
I would prefer to make the new deserializer default in next release. by that time, I assume atleast you would have tested it and would be running it in prod and so we know it works smoothly end to end.

sbernauer replied:
Alrigth, seems perfectly fine to me. Fixed

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/AvroKafkaSource.java:
```diff
@@ -52,6 +54,10 @@
   private static final String NATIVE_KAFKA_VALUE_DESERIALIZER_PROP = "value.deserializer";
+  // These are settings used to pass things to KafkaAvroDeserializer
+  public static final String KAFKA_AVRO_VALUE_DESERIALIZER_SCHEMA = "hoodie.deltastreamer.source.kafka.value.deserializer.schema";
+  public static final String KAFKA_AVRO_VALUE_DESERIALIZER_PROPERTY_PREFIX = "hoodie.deltastreamer.source.kafka.value.deserializer.";
```
can we move this ahead of line 58 and use this to form  the other variables. 
for eg:
```
public static final String KAFKA_AVRO_VALUE_DESERIALIZER_SCHEMA = KAFKA_AVRO_VALUE_DESERIALIZER_PROPERTY_PREFIX + "schema";
```

sbernauer replied:
Thanks for the suggestion, fixed

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/AvroKafkaSource.java:
```diff
@@ -60,7 +66,7 @@ public AvroKafkaSource(TypedProperties props, JavaSparkContext sparkContext, Spa
 
     props.put(NATIVE_KAFKA_KEY_DESERIALIZER_PROP, StringDeserializer.class);
-    String deserializerClassName = props.getString(DataSourceWriteOptions.KAFKA_AVRO_VALUE_DESERIALIZER(), "");
+    String deserializerClassName = props.getString(KAFKA_AVRO_VALUE_DESERIALIZER_OPT_KEY(), DEFAULT_KAFKA_AVRO_VALUE_DESERIALIZER_OPT_VAL());
```
we need to be careful in line 71. since we have updated the default value, we may never go into if condition of line 71. Can you fix that please.

sbernauer replied:
Hi, i would prefer to leave the block in - even if we turn the feature on by default. Doing so we have the possibility to turn the feature explicitly off (`hoodie.deltastreamer.source.kafka.value.deserializer.class=`) if it causes any problems

nsivabalan replied:
I am not sure I get you. Can you clarify please. 
As of now, it looks fine, since the default value is empty string. I am talking about when we make the new deser class as default. We can come back to revisit this when we fix the default value.

sbernauer replied:
I agree with you.
But having a way to disable the feature (setting `(hoodie.deltastreamer.source.kafka.value.deserializer.class` to empty) is always a good idea. But for this to work we need to leave in the Codeblock checking for an empty string.
So I would suggest to leave it in regardless of the default value.

sbernauer replied:
While rebasing i found a way how we can remove this isEmpty() check. Take a look at the current patch.

nsivabalan commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -381,7 +381,6 @@ object DataSourceWriteOptions {
-  val KAFKA_AVRO_VALUE_DESERIALIZER = "hoodie.deltastreamer.source.kafka.value.deserializer.class"
-  // Schema to be used in custom kakfa deserializer
-  val KAFKA_AVRO_VALUE_DESERIALIZER_SCHEMA = "hoodie.deltastreamer.source.kafka.value.deserializer.schema"
+  val KAFKA_AVRO_VALUE_DESERIALIZER_OPT_KEY = "hoodie.deltastreamer.source.kafka.value.deserializer.class"
```
nit: can we reuse the property prefix variable here as well? KAFKA_AVRO_VALUE_DESERIALIZER_PROPERTY_PREFIX

sbernauer replied:
I also thought of that but I was unable to import the hudi-utilities Class in the hudi-spark module I think. It would be great if you could give a hint how to achieve this.

sbernauer replied:
Just tried it out again. Adding hudi-utilities as dependency to the pom of hudi-spark introduces a circular dependency :/

nsivabalan reviewed (commented):
Looks like you need some rebasing. Can you please check it out. 
@vinothchandar : Do we have a guidelines or link for adding new configs. I think we need to do something extra so that configs are auto generated right.

sbernauer commented:
Rebase done

sbernauer commented:
I should have addressed all feedback. We have tested this patch again and have rolled it out to production.
Can you please have a look if everything is fine?

vinothchandar reviewed (approved):
LGTM. 

reg, auto generation of configs. HUDI-1241 is tracking it.

vinothchandar commented:
@hudi-bot run azure

sbernauer commented:
Thanks @vinothchandar!
As the patch uses the ConfigProperty class the generation of the docs should work fine as my understanding

sbernauer commented:
Thanks a lot for merging!
